### PR TITLE
Add multiplex operator

### LINKED ISSUE #4010: Multiplex Operator

### PR DESCRIPTION
**Pull request by kuke**
Resolve #4010

### CODE REVIEW

File: paddle/operators/multiplex_op.cc [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    PADDLE_ENFORCE(!ctx.MultiInputVar("X").empty(),
+                   "Input(X) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Out"),
+                            "Output(Out) shouldn't be null.");
+    auto ins = ctx.MultiInput<Tensor>("X");
+    auto *out = ctx.Output<LoDTensor>("Out");
+    auto num_ins = ins.size();
+    PADDLE_ENFORCE(num_ins > 2,
+                   "multiplex operator should have more than 2 inputs.");
+    PADDLE_ENFORCE_EQ(ins[0]->dims().size(), 1,
```
Comment (QiJune) [inline paddle/operators/multiplex_op.cc]: We also have to check the index in ins[0], index in ins[0] must less than ins[0]->dims()

File: paddle/operators/multiplex_op.cc [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE(!ctx.MultiInputVar("X").empty(),
+                   "Input(X) should not be null");
+    PADDLE_ENFORCE(!ctx.MultiOutputVar(framework::GradVarName("X")).empty(),
+                   "Output(X@Grad) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) shouldn't be null.");
+    auto d_ins = ctx.MultiOutput<LoDTensor>(framework::GradVarName("X"));
+    auto ins = ctx.MultiInput<Tensor>("X");
+    // don;t compute gradient for index
```
Comment (QiJune) [inline paddle/operators/multiplex_op.cc]: don;t  --> don't

File: paddle/operators/multiplex_op.cu [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    auto rows = ins[1]->dims()[0];
+    auto cols = ins[1]->dims()[1];
+    // copy index to cpu
+    Tensor index_t_cpu;
+    index_t_cpu.CopyFrom<T>(*(ins[0]), paddle::platform::CPUPlace());
+    auto index = index_t_cpu.data<T>();
+    for (auto i = 0; i < rows; i++) {
+      int k = (int)index[i] + 1;
+      cudaMemcpy(out->data<T>() + i * cols, ins[k]->data<T>() + i * cols,
```
Comment (QiJune) [inline paddle/operators/multiplex_op.cu]: Please use  cuda stream.
```
auto stream = reinterpret_cast<const platform::CUDADeviceContext&>(
                       ctx.device_context())
                       .stream();
platform::GPUPlace place = boost::Get<platform::GPUPlace>(ctx.GetPlace());
memory::Copy(place, out->data<T>() + i * cols, place, ins[k]->data<T>() + i * cols, cols * sizeof(T), stream);

```

File: paddle/operators/multiplex_op.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    auto ins = ctx.MultiInput<framework::Tensor>("X");
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+    out->mutable_data<T>(ctx.GetPlace());
+
+    auto index = ins[0]->data<T>();
+    auto rows = ins[1]->dims()[0];
+    auto cols = ins[1]->dims()[1];
+    for (auto i = 0; i < rows; i++) {
+      int k = (int)index[i] + 1;
+      memcpy(out->data<T>() + i * cols, ins[k]->data<T>() + i * cols,
```
Comment (QiJune) [inline paddle/operators/multiplex_op.h]: Maybe we can combine cpu code and cuda code in one file.
```
template <typename Place, typename T>
class MultiplexKernel : public framework::OpKernel 
```
We can use 
```
t.device(context.GetEigenDevice<Place>()) = t.constant(static_cast<T>(0));
```
for set cpu/gpu to zero

And we can use 
```
memory::Copy
```
 for both cpu/gpu copy

File: paddle/operators/multiplex_op.cc (hunk 1 above)
Comment (kuke) [reply paddle/operators/multiplex_op.cc]: Done. Add the index check in the forward compute function.

File: paddle/operators/multiplex_op.cc [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using LoDTensor = framework::LoDTensor;
+
+class MultiplexOp : public framework::OperatorWithKernel {
+ public:
+  MultiplexOp(const std::string &type, const framework::VariableNameMap &inputs,
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/multiplex_op.cc]: Modified

File: paddle/operators/multiplex_op.cc (hunk 2 above)
Comment (kuke) [reply paddle/operators/multiplex_op.cc]: Done

File: paddle/operators/multiplex_op.cu (hunk 3 above)
Comment (kuke) [reply paddle/operators/multiplex_op.cu]: Done

File: paddle/operators/multiplex_op.h (hunk 4 above)
Comment (kuke) [reply paddle/operators/multiplex_op.h]: Done

Comment (QiJune) [reply paddle/operators/multiplex_op.h]: It seems that merge CPU/GPU code together is not a good idea here. I make a mistake.
If CPU and GPU both use Eigen, we can reuse codes easily. But if not, it's actually better to split CPU and GPU implementation.

Comment (kuke) [reply paddle/operators/multiplex_op.h]: Done. split CPU/GPU code again.

**Pull request merged**